src/engine/Preprocessor.cpp

```diff
@@ -781,6 +794,278 @@ void Preprocessor::dumpAllBounds( const String &message )
+    */
+    struct Index
+    {
+        Index()
+            : _layer( 0 )
+            , _neuron( 0 )
+        {
+        }
+
+        Index( unsigned layer, unsigned neuron )
+            : _layer( layer )
+            , _neuron( neuron )
+        {
+        }
+
+        bool operator<( const Index &other ) const
+        {
+            if ( _layer < other._layer )
+                return true;
+            if ( _layer > other._layer )
+                return false;
+
+            return _neuron < other._neuron;
+        }
+
+        unsigned _layer;
+        unsigned _neuron;
+    };
+
+    struct NeuronInformation
+    {
+        NeuronInformation()
+            : _weightedSumEquation( NULL )
+            , _activationFunction( NULL )
+        {
+        }
+
+        unsigned _weightedSumVariable;
+        unsigned _activationVariable;
+
+        PiecewiseLinearFunctionType _activationType;
+
+        const Equation *_weightedSumEquation;
+        const PiecewiseLinearConstraint *_activationFunction;
+    };
+
+    Map<Index, NeuronInformation> indexToNeuron;
+    Map<unsigned, Index> variableToIndex;
+    Map<unsigned, unsigned> layerToLayerSize;
+
+    ASSERT( !_preprocessed._networkLevelReasoner );
+
+    const List<Equation> &equations( _preprocessed.getEquations() );
+    const List<PiecewiseLinearConstraint *> plConstraints( _preprocessed.getPiecewiseLinearConstraints() );
+
+    // If the activation functions are not yet supported by the NLR,
+    // abort
+    for ( const auto constraint : plConstraints )
+        if ( !NetworkLevelReasoner::functionTypeSupported( constraint->getType() ) )
+            return false;
+
+    // Attempt to figure out the layered structure
+    List<unsigned> inputLayer;
+    List<unsigned> previousLayer;
+    List<unsigned> currentLayer;
+    List<unsigned> outputLayer;
+    unsigned currentLayerIndex = 1;
+
+    // The input and output layer should be defined in the input query
+    inputLayer = _preprocessed.getInputVariables();
+    previousLayer = inputLayer;
+    unsigned numVariablesHandledSoFar = inputLayer.size();
+    outputLayer = _preprocessed.getOutputVariables();
+    layerToLayerSize[0] = inputLayer.size();
+
+    // Store the input neurons in the topology
+    unsigned count = 0;
+    for ( const auto &neuron : inputLayer )
+    {
+        NeuronInformation neuronInfo;
+        neuronInfo._activationVariable = neuron;
+        Index index( 0, count );
+        indexToNeuron[index] = neuronInfo;
+        variableToIndex[neuron] = index;
+        ++count;
+    }
+
+    // Now, attempt to figure out the topology, layer by layer
+    while ( true )
+    {
+        currentLayer.clear();
+        for ( const auto &eq : equations )
+        {
+            // Only consider equalities
+            if ( eq._type != Equation::EQ )
+                continue;
+
+            Set<unsigned> eqVars = eq.getParticipatingVariables();
+            for ( const auto &var : previousLayer )
```

Comment:
In case of variables skipping layers, even though less efficient, we could go through all the already processed variables? There is the issue of indexing, but least layer-depth indexing is probably a reasonable normal form for such nodes.

Reply:
I agree that layer skipping is a nice feature. We should probably add it in the future. However, currently the SBT mechanism would break if we did that, so this is a topic for a separate PR.